### ISSUE TITLE
Support for KeyFactory Interfaces

### ISSUE THREAD

**Issue by huyngo1216** [corretto/amazon-corretto-crypto-provider#90]
We are working on a project to benchmark improvements that can be achieved by using ACCP that currently uses BouncyCastle. However, the `KeyFactory` algorithm types are not yet supported by ACCP. Specifically, the "RSA" algorithm would be most helpful for our use case. Do you have plans to support that in the future?

For example, BouncyCastle exposes this utility to convert keys. https://www.bouncycastle.org/docs/pkixdocs1.5on/org/bouncycastle/openssl/jcajce/JcaPEMKeyConverter.html

**Comment by SalusaSecondus** [comment]
Currently ACCP uses whatever `KeyFactory` implementations are already configured in your application. Up till now this has been sufficient for the use-cases we've encountered.

What, specifically, are you trying to do? The linked code could still be used (without explicitly setting the provider) to convert to standard `java.security.PrivateKey` and `java.security.PublicKey` which would then work correctly with ACCP.

**Comment by huyngo1216** [comment]
Ah, thank you for the clarification. I misunderstood that the implementations can be intermixed within the application. I'll close the issue for now.

**Comment by SalusaSecondus** [comment]
Not a problem. We'd love to see your results. Please be aware of the following (interesting) factors:
* The three different ways of getting an implementation have very different performance characteristics (regardless of provider):
  * Highest priority implementation (no provider requested)
  * Named provider (explicit provider by String name)
  * Explicit provider (provider object passed in)
* Creating multiple copies of a provider is an anti-pattern and can cause performance problems.
* The `getInstance` call be be expensive in Java (and differently expensive for different providers. However, the cost is sometimes deferred till the `.init` or `.getProvider` call.
* Some providers (such as ACCP) have optimized code-paths for key reuse. So, if you use the same object (`Cipher` or `Signature` in your case) with the same key multiple times in a row (with or without re-initializing), they will remember state to decrease cost of future use. (Please note that the `PublicKey` and the `PrivateKey` are different keys and ACCP cannot optimize use across them.)

**Comment by huyngo1216** [comment]
Dropping in the library and running against some benchmarking we have in place, we're seeing a ~25% latency + throughput improvement!

Following up on your second point:
> Creating multiple copies of a provider is an anti-pattern and can cause performance problems

Is there any reference documentation/blog posts you can point me to so I can read up on that?

**Comment by SalusaSecondus** [comment]
I don't have any references from that, just personal experience.

I found some code which kept creating new instances of `BouncyCastleProvider`. That's an expensive object to create and resulted in extra GC pressure as the old ones got discarded.

But 25%? That's great! Things like this are exactly why we released this library in the first place and it's nice to hear that it is helping out others.
